Metadata/MetadataConverter.php

```diff
@@ -37,49 +37,56 @@ public function convert(ClassHierarchyMetadata $metadata)
         $definitions = array();
 
         $previous = null;
+        /** @var ClassMetadata $classMetadata */
         foreach ($metadata->classMetadata as $classMetadata) {
-            if (null === $previous && null === $classMetadata->parent) {
-                $definition = new Definition();
-            } else {
-                $definition = new DefinitionDecorator(
-                    $classMetadata->parent ?: $previous->id
-                );
-            }
+            foreach ($classMetadata->getServices() as $service) {
+                if (null === $previous && null === @$service['parent']) {
+                    $definition = new Definition();
+                } else {
+                    $definition = new DefinitionDecorator(
+                        @$service['parent'] ?: $previous->id
+                    );
+                }
 
-            $definition->setClass($classMetadata->name);
-            if (null !== $classMetadata->scope) {
-                $definition->setScope($classMetadata->scope);
-            }
-            if (null !== $classMetadata->public) {
-                $definition->setPublic($classMetadata->public);
-            }
-            if (null !== $classMetadata->abstract) {
-                $definition->setAbstract($classMetadata->abstract);
-            }
-            if (null !== $classMetadata->arguments) {
-                $definition->setArguments($classMetadata->arguments);
-            }
+                $definition->setClass($classMetadata->name);
+                if (null !== @$service['scope']) {
+                    $definition->setScope(@$service['scope']);
+                }
+                if (null !== @$service['public']) {
+                    $definition->setPublic(@$service['public']);
+                }
+                if (null !== @$service['abstract']) {
+                    $definition->setAbstract(@$service['abstract']);
+                }
+                if (null !== $classMetadata->arguments) {
+                    $definition->setArguments($classMetadata->arguments);
+                }
 
-            $definition->setMethodCalls($classMetadata->methodCalls);
-            $definition->setTags($classMetadata->tags);
-            $definition->setProperties($classMetadata->properties);
+                $definition->setMethodCalls($classMetadata->methodCalls);
+                $definition->setTags($classMetadata->tags);
+                $definition->setProperties($classMetadata->properties);
 
-            if (null === $classMetadata->id) {
-                $classMetadata->id = '_jms_di_extra.unnamed.service_'.$count++;
-            }
+                if (null === @$service['id']) {
+                    $classMetadata->id = '_jms_di_extra.unnamed.service_' . $count++;
```

Comment:
It looks strange to me to use this ``ClassMetadata::$id`` as it is now deprecated...